mdata/aggmetric.go

```diff
@@ -137,27 +137,33 @@ func (a *AggMetric) getChunk(pos int) *chunk.Chunk {
 	return a.Chunks[pos]
 }
 
-func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSpan, from, to uint32) Result {
+func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSpan, from, to uint32) (Result, error) {
 	// no lock needed cause aggregators don't change at runtime
 	for _, a := range a.aggregators {
 		if a.span == aggSpan {
+			var agg *AggMetric
 			switch consolidator {
 			case consolidation.None:
 				panic("cannot get an archive for no consolidation")
 			case consolidation.Avg:
 				panic("avg consolidator has no matching Archive(). you need sum and cnt")
 			case consolidation.Cnt:
-				return a.cntMetric.Get(from, to)
+				agg = a.cntMetric
 			case consolidation.Lst:
-				return a.lstMetric.Get(from, to)
+				agg = a.lstMetric
 			case consolidation.Min:
-				return a.minMetric.Get(from, to)
+				agg = a.minMetric
 			case consolidation.Max:
-				return a.maxMetric.Get(from, to)
+				agg = a.maxMetric
 			case consolidation.Sum:
-				return a.sumMetric.Get(from, to)
+				agg = a.sumMetric
+			default:
+				panic(fmt.Sprintf("AggMetric.GetAggregated(): unknown consolidator %q", consolidator))
```

Comment:
Shouldn't this case be removed?

Reply:
Not necessarily. While it can be handled below, it is a valid (and separate) issue that unexpected values got passed in here.

Reply:
Aha, I wasn't thinking about `agg` getting set to a nil value above.  In any case, I'd be very keen to see the `panic` calls converted to error returns unless there's a compelling reason not to.

Reply:
That's fair. I wasn't sure if panics got treated differently than errors, but (aside from runtime error panics) they both get turned into `500` responses. Ideally, a `400` would return from some of these.

Reply:
note: the "automatic turning panics into http 500 responses" only works when the recover happens (the macaron middleware) in the same goroutine as the panic. in this case not, because we spawn new goroutines for all s.getTargetsLocal calls. (there's one or two other open tickets describing running into this.. unpleasantery. needless to say, significant problem that we need to resolve)

regarding when (not) to panic, we have a policy for that now : https://github.com/grafana/metrictank/blob/master/docs/CONTRIBUTING#L19
if the articulation needs revising, please do tell :)

for this PR, you don't need to worry about cleaning up panics that were already there, we can do that later. (though you're welcome to)

Reply:
I switched the panics to errors, but I'm not sure at what granularity we need the metrics to be posted. There could be a metric per, or just a generic "GetAggregated" error. I'm also not sure of the naming convention, since the memory index does "idx.memory" and ccache does "cache.metric"